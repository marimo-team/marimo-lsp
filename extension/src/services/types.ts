import type { Brand } from "effect";
import type * as vscode from "vscode";

export type MarimoNotebookDocument = Brand.Branded<
  vscode.NotebookDocument,
  "MarimoNotebookDocument"
>;
