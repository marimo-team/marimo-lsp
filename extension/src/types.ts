import type { components as Api } from "@marimo-team/openapi/src/api";
import { Brand } from "effect";
import type * as vscode from "vscode";
import type * as lsp from "vscode-languageclient/node";
import type { MarimoNotebook } from "./schemas.ts";

export type { CellRuntimeState } from "@marimo-team/frontend/unstable_internal/core/cells/types.ts";

type Schemas = Api["schemas"];

export type MessageOperation = Schemas["KnownUnions"]["operation"];
export type MessageOperationOf<T extends MessageOperation["op"]> = Extract<
  MessageOperation,
  { op: T }
>;
export type CellMessage = MessageOperationOf<"cell-op">;
export type VariableValuesOp = MessageOperationOf<"variable-values">;
export type VariablesOp = MessageOperationOf<"variables">;

export type NotebookUri = Brand.Branded<string, "NotebookUri">;

// Only way to get our NotebookUri type is from the server or a vscode.NotebookDocument
const NotebookUri = Brand.nominal<NotebookUri>();
export function getNotebookUri(doc: vscode.NotebookDocument): NotebookUri {
  return NotebookUri(doc.uri.toString());
}

interface NotebookScoped<T> {
  notebookUri: NotebookUri;
  inner: T;
}

interface SessionScoped<T> extends NotebookScoped<T> {
  executable: string;
}

type RunRequest = Schemas["RunRequest"];
type SetUIElementValueRequest = Schemas["SetUIElementValueRequest"];
type FunctionCallRequest = Schemas["FunctionCallRequest"];
interface DeserializeRequest {
  source: string;
}
interface SerializeRequest {
  notebook: typeof MarimoNotebook.Type;
}
interface DebugAdapterRequest {
  sessionId: string;
  message: vscode.DebugProtocolMessage;
}
// biome-ignore lint/complexity/noBannedTypes: We need this for over the wire
type InterruptRequest = {};

// client -> language server
type MarimoCommandMap = {
  "marimo.run": SessionScoped<RunRequest>;
  "marimo.set_ui_element_value": NotebookScoped<SetUIElementValueRequest>;
  "marimo.function_call_request": NotebookScoped<FunctionCallRequest>;
  "marimo.dap": NotebookScoped<DebugAdapterRequest>;
  "marimo.interrupt": NotebookScoped<InterruptRequest>;
  "marimo.serialize": SerializeRequest;
  "marimo.deserialize": DeserializeRequest;
};

type MarimoCommandMessageOf<K extends keyof MarimoCommandMap> = {
  [C in keyof MarimoCommandMap]: {
    command: C;
    params: MarimoCommandMap[C];
  };
}[K];

/** Subset of commands allowed to be dispatched by the renderer */
type RendererCommandMap = {
  [K in
    | "marimo.set_ui_element_value"
    | "marimo.function_call_request"]: MarimoCommandMap[K]["inner"];
};
type RendererCommandMessageOf<K extends keyof RendererCommandMap> = {
  [C in keyof RendererCommandMap]: {
    command: C;
    params: RendererCommandMap[C];
  };
}[K];

export type MarimoCommand = MarimoCommandMessageOf<keyof MarimoCommandMap>;

// renderer -> extension
export type RendererCommand = RendererCommandMessageOf<
  keyof RendererCommandMap
>;

// extension -> renderer
export type RendererReceiveMessage =
  | MessageOperationOf<"remove-ui-elements">
  | MessageOperationOf<"send-ui-element-message">
  | MessageOperationOf<"function-call-result">;

// Language server -> client
type MarimoNotificationMap = {
  "marimo/operation": { notebookUri: NotebookUri; operation: MessageOperation };
  "marimo/dap": { sessionId: string; message: vscode.DebugProtocolMessage };
  "window/logMessage": lsp.LogMessageParams;
};
export type MarimoNotification = keyof MarimoNotificationMap;
export type MarimoNotificationOf<K extends MarimoNotification> = {
  [C in MarimoNotification]: MarimoNotificationMap[C];
}[K];
