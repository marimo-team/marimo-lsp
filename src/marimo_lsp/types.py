"""Types for marimo-lsp."""

from __future__ import annotations

from typing import TYPE_CHECKING

import msgspec
from marimo._messaging.msgspec_encoder import encode_json_str

if TYPE_CHECKING:
    from marimo._config.config import MarimoConfig
    from marimo._runtime.requests import AppMetadata


def encode_kernel_args(args: KernelArgs) -> str:
    """Encode kernel args as JSON."""
    return encode_json_str(args)


def decode_kernel_args(json: str) -> KernelArgs:
    """Encode kernel args as JSON."""
    return msgspec.json.decode(json, type=KernelArgs)


class KernelArgs(msgspec.Struct):
    """Args to send to the kernel."""

    configs: dict
    app_metadata: AppMetadata
    user_config: MarimoConfig
    log_level: int
