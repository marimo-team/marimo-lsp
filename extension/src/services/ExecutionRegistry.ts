import {
  type Brand,
  Data,
  Effect,
  FiberSet,
  HashMap,
  Option,
  Ref,
  String,
} from "effect";
import type * as vscode from "vscode";
import { assert } from "../assert.ts";
import {
  type CellRuntimeState,
  createCellRuntimeState,
  transitionCell,
} from "../utils/transitionCell.ts";
import { type CellMessage, getNotebookUri } from "../types.ts";
import { CellStateManager } from "./CellStateManager.ts";
import type { NotebookController } from "./NotebookControllerFactory.ts";
import { VsCode } from "./VsCode.ts";

export class ExecutionRegistry extends Effect.Service<ExecutionRegistry>()(
  "ExecutionRegistry",
  {
    dependencies: [CellStateManager.Default],
    scoped: Effect.gen(function* () {
      const code = yield* VsCode;
      const cellStateManager = yield* CellStateManager;
      const ref = yield* Ref.make(HashMap.empty<NotebookCellId, CellEntry>());

      yield* Effect.addFinalizer(() =>
        Ref.update(ref, (map) => {
          HashMap.forEach(map, (entry) =>
            Option.map(entry.pendingExecution, (e) => e.end(true)),
          );
          return HashMap.empty();
        }),
      );
      const runFork = yield* FiberSet.makeRuntime();
      return {
        handleInterrupted(editor: vscode.NotebookEditor) {
          return Ref.update(ref, (map) =>
            HashMap.map(map, (cell) =>
              Option.match(cell.pendingExecution, {
                onSome: () => cell.editor === editor,
                onNone: () => false,
              })
                ? CellEntry.interrupt(cell)
                : cell,
            ),
          );
        },
        handleCellOperation: Effect.fnUntraced(function* (
          msg: CellMessage,
          deps: {
            editor: vscode.NotebookEditor;
            controller: NotebookController;
          },
        ) {
          const { editor, controller } = deps;
          const cellId = extractCellId(msg);

          const cell = yield* Ref.modify(ref, (map) => {
            const prev = Option.match(HashMap.get(map, cellId), {
              onSome: (cell) => cell,
              onNone: () => CellEntry.make(cellId, deps.editor),
            });
            const update = CellEntry.transition(prev, msg);
            return [update, HashMap.set(map, cellId, update)];
          });

          switch (msg.status) {
            case "queued": {
              // Clear stale state when cell is queued for execution
              const notebookUri = getNotebookUri(editor.notebook);
              const notebookCell = getNotebookCell(editor.notebook, cell.id);
              yield* cellStateManager.clearCellStale(
                notebookUri,
                notebookCell.index,
              );

              yield* Ref.update(ref, (map) => {
                const handle = HashMap.get(map, cellId).pipe(
                  Option.andThen((v) => v.pendingExecution),
                );
                if (Option.isSome(handle)) {
                  // Need to clear existing
                  handle.value.end(true);
                }
                const update = CellEntry.withExecution(
                  cell,
                  new ExecutionHandle({
                    runId: Option.getOrThrow(extractRunId(msg)),
                    inner: controller.createNotebookCellExecution(notebookCell),
                  }),
                );
                return HashMap.set(map, cellId, update);
              });
              yield* Effect.logDebug("Cell queued for execution").pipe(
                Effect.annotateLogs({ cellId }),
              );
              return;
            }

            case "running": {
              if (Option.isNone(cell.pendingExecution)) {
                yield* Effect.logWarning(
                  "Got running message but no cell execution found.",
                );
              }
              const update = yield* Ref.modify(ref, (map) => {
                const update = CellEntry.start(cell, msg.timestamp ?? 0);
                return [update, HashMap.set(map, cellId, update)];
              });
              yield* Effect.logDebug("Cell execution started").pipe(
                Effect.annotateLogs({ cellId }),
              );
              yield* CellEntry.maybeUpdateCellOutput(update, code);
              return;
            }

            case "idle": {
              if (Option.isNone(cell.pendingExecution)) {
                yield* Effect.logWarning(
                  "Got idle message but no cell execution found.",
                );
              }
              // MUST modify cell output before `ExecutionHandle.end`
              yield* CellEntry.maybeUpdateCellOutput(cell, code);

              {
                // FIXME: stdin/stdout are flushed every 10ms, so wait 50ms
                // to ensure all related events arrive before finalizing.
                //
                // marimo doesn't set a run_id for idle messages, so we can't compare
                // against the incoming message to detect if a new execution has started.
                //
                // Ref: https://github.com/marimo-team/marimo/blob/3644b6f/marimo/_messaging/ops.py#L148-L151
                //
                // Instead, we capture the `lastRunId` before the timeout and compare it
                // when finalizing. If a new execution starts before the timeout fires,
                // the `lastRunId` will have changed and we skip finalization.
                const lastRunId = cell.lastRunId;
                const finalize = Ref.update(ref, (map) => {
                  const fresh = HashMap.get(map, cellId);

                  if (Option.isNone(fresh)) {
                    return map;
                  }

                  const isDifferentRun = !Option.getEquivalence(
                    String.Equivalence,
                  )(fresh.value.lastRunId, lastRunId);

                  if (isDifferentRun) {
                    return map;
                  }

                  return HashMap.set(
                    map,
                    cellId,
                    CellEntry.end(fresh.value, true, msg.timestamp),
                  );
                });

                setTimeout(() => runFork(finalize), 50);
              }

              yield* Effect.logDebug("Cell execution completed").pipe(
                Effect.annotateLogs({ cellId }),
              );
              return;
            }

            default: {
              yield* CellEntry.maybeUpdateCellOutput(cell, code);
            }
          }
        }),
      };
    }),
  },
) {}

class ExecutionHandle extends Data.TaggedClass("ExecutionHandle")<{
  readonly inner: vscode.NotebookCellExecution;
  readonly runId: RunId;
}> {
  start(startTime: number) {
    return this.inner.start(startTime);
  }
  end(success: boolean, endTime?: number) {
    return this.inner.end(success, endTime);
  }
  static updateCellOutput = Effect.fnUntraced(function* (
    execution: ExecutionHandle,
    options: {
      cellId: NotebookCellId;
      state: CellRuntimeState;
      code: VsCode;
    },
  ) {
    const { cellId, state, code } = options;
    yield* Effect.tryPromise(() =>
      execution.inner.replaceOutput(
        new code.NotebookCellOutput([
          code.NotebookCellOutputItem.json(
            { cellId, state },
            "application/vnd.marimo.ui+json",
          ),
        ]),
      ),
    ).pipe(
      Effect.catchAllCause((cause) =>
        Effect.logError("Failed to update cell output", cause).pipe(
          Effect.annotateLogs({ cellId }),
        ),
      ),
    );
  });
}

class CellEntry extends Data.TaggedClass("CellEntry")<{
  readonly id: NotebookCellId;
  readonly state: CellRuntimeState;
  readonly editor: vscode.NotebookEditor;
  readonly pendingExecution: Option.Option<ExecutionHandle>;
  readonly lastRunId: Option.Option<RunId>;
}> {
  static make(id: NotebookCellId, editor: vscode.NotebookEditor) {
    return new CellEntry({
      id,
      editor,
      state: createCellRuntimeState(),
      pendingExecution: Option.none(),
      lastRunId: Option.none(),
    });
  }
  static transition(cell: CellEntry, message: CellMessage) {
    return new CellEntry({
      ...cell,
      state: transitionCell(cell.state, message),
    });
  }
  static withExecution(cell: CellEntry, execution: ExecutionHandle) {
    return new CellEntry({
      ...cell,
      lastRunId: Option.some(execution.runId),
      pendingExecution: Option.some(execution),
    });
  }
  static interrupt(cell: CellEntry) {
    if (Option.isSome(cell.pendingExecution)) {
      cell.pendingExecution.value.end(false);
    }
    return new CellEntry({
      ...cell,
      pendingExecution: Option.none(),
    });
  }
  static start(cell: CellEntry, timestamp: number) {
    if (Option.isSome(cell.pendingExecution)) {
      cell.pendingExecution.value.inner.start(timestamp * 1000);
    }
    return new CellEntry({ ...cell });
  }
  static end(cell: CellEntry, success: boolean, timestamp?: number) {
    if (Option.isSome(cell.pendingExecution)) {
      cell.pendingExecution.value.inner.end(
        success,
        timestamp ? timestamp * 1000 : undefined,
      );
    }
    return new CellEntry({
      ...cell,
      pendingExecution: Option.none(),
    });
  }
  static maybeUpdateCellOutput = Effect.fnUntraced(function* (
    cell: CellEntry,
    code: VsCode,
  ) {
    const { pendingExecution, id: cellId, state } = cell;
    if (Option.isNone(pendingExecution)) {
      yield* Effect.logWarning("No pending execution to update.");
      return;
    }
    yield* ExecutionHandle.updateCellOutput(pendingExecution.value, {
      cellId,
      state,
      code,
    }).pipe(
      Effect.catchAllCause((cause) =>
        Effect.logError("Failed to update cell output", cause).pipe(
          Effect.annotateLogs({ cellId }),
        ),
      ),
    );
  });
}

type RunId = Brand.Branded<string, "RunId">;
function extractRunId(msg: CellMessage) {
  return Option.fromNullable(msg.run_id) as Option.Option<RunId>;
}

type NotebookCellId = Brand.Branded<string, "CellId">;
function extractCellId(msg: CellMessage) {
  return msg.cell_id as NotebookCellId;
}

function getNotebookCell(
  notebook: vscode.NotebookDocument,
  cellId: NotebookCellId,
): vscode.NotebookCell {
  const cell = notebook
    .getCells()
    .find((c) => c.document.uri.toString() === cellId);
  assert(cell, `No cell id ${cellId} in notebook ${notebook.uri.toString()} `);
  return cell;
}
