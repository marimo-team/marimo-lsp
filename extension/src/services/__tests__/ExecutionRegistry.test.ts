import { expect, it } from "@effect/vitest";
import { createCellRuntimeState } from "@marimo-team/frontend/unstable_internal/core/cells/types.ts";
import { Effect } from "effect";
import { TestVsCodeLive } from "../../__mocks__/TestVsCode.ts";
import type { CellRuntimeState } from "../../types.ts";
import { VsCode } from "../VsCode.ts";
import { buildCellOutputs, NotebookCellId } from "../ExecutionRegistry.ts";

const ExecutionRegistryTestLive = TestVsCodeLive;

const CELL_ID = "test-cell-id" as NotebookCellId;

// Convert Uint8Array data to strings for readable snapshots
function normalizeOutputsForSnapshot(outputs: unknown) {
  if (!Array.isArray(outputs)) return outputs;

  return outputs.map((output: any) => {
    if (!output.items) return output;

    return {
      ...output,
      items: output.items.map((item: any) => {
        if (item.data instanceof Uint8Array) {
          const decoder = new TextDecoder();
          return {
            ...item,
            data: decoder.decode(item.data),
          };
        }
        return item;
      }),
    };
  });
}

it.layer(ExecutionRegistryTestLive)("buildCellOutputs", (it) => {
  it.effect(
    "handles stdout output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Hello from stdout",
            timestamp: 0,
          },
        ],
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles stderr output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Error message",
            timestamp: 0,
          },
        ],
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles multiple console outputs",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Line 1\n",
            timestamp: 0,
          },
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Line 2\n",
            timestamp: 1,
          },
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Warning: something happened\n",
            timestamp: 2,
          },
        ],
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles marimo error output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        output: {
          mimetype: "application/vnd.marimo+error",
          channel: "marimo-error",
          data: [
            {
              type: "syntax",
              msg: "Invalid syntax",
              cell_id: CELL_ID.toString(),
            },
          ],
          timestamp: 0,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles HTML output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        output: {
          mimetype: "text/html",
          channel: "output",
          data: "<div>Hello <b>world</b></div>",
          timestamp: 0,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles JSON output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        output: {
          mimetype: "application/json",
          channel: "output",
          data: { foo: "bar", count: 42 },
          timestamp: 0,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles mixed output and console streams",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Processing...\n",
            timestamp: 0,
          },
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Warning: deprecated function\n",
            timestamp: 1,
          },
        ],
        output: {
          mimetype: "text/html",
          channel: "output",
          data: "<div>Result: 42</div>",
          timestamp: 2,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles stdin output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stdin",
            data: "Enter your name: ",
            timestamp: 0,
          },
        ],
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles empty state",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = createCellRuntimeState();

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      // Should still have the marimo UI output
      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles multiple errors in marimo error output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        output: {
          mimetype: "application/vnd.marimo+error",
          channel: "marimo-error",
          data: [
            {
              type: "exception",
              msg: "ValueError: invalid value",
              exception_type: "ValueError",
            },
            {
              type: "ancestor-stopped",
              msg: "Ancestor cell was stopped",
              raising_cell: "other-cell-id",
            },
          ],
          timestamp: 0,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles multiple stderr errors",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Error 1: Connection failed\n",
            timestamp: 0,
          },
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Error 2: Retry failed\n",
            timestamp: 1,
          },
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Error 3: Timeout\n",
            timestamp: 2,
          },
        ],
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles stdout + stderr + output together",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        consoleOutputs: [
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Starting computation...\n",
            timestamp: 0,
          },
          {
            mimetype: "text/plain",
            channel: "stderr",
            data: "Warning: using deprecated API\n",
            timestamp: 1,
          },
          {
            mimetype: "text/plain",
            channel: "stdout",
            data: "Computation complete\n",
            timestamp: 2,
          },
        ],
        output: {
          mimetype: "application/json",
          channel: "output",
          data: { result: "success", value: 100 },
          timestamp: 3,
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );

  it.effect(
    "handles application/vnd.marimo+traceback output",
    Effect.fnUntraced(function* () {
      const code = yield* VsCode;
      const state: CellRuntimeState = {
        ...createCellRuntimeState(),
        output: {
          mimetype: "application/vnd.marimo+traceback",
          channel: "output",
          data: "<b>Traceback (most recent call last):</b>\n  File \"<stdin>\", line 1, in <module>\nTypeError: invalid value",
        },
      };

      const outputs = buildCellOutputs(
        CELL_ID,
        state,
        code,
      );

      expect(normalizeOutputsForSnapshot(outputs)).toMatchSnapshot();
    }),
  );
});

