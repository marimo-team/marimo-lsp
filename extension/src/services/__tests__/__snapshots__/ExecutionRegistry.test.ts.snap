// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildCellOutputs > filters out empty text/plain stderr output 1`] = `[]`;

exports[`buildCellOutputs > filters out empty text/plain stdout output 1`] = `[]`;

exports[`buildCellOutputs > filters out empty traceback output 1`] = `[]`;

exports[`buildCellOutputs > filters out null output data 1`] = `[]`;

exports[`buildCellOutputs > filters out undefined output data 1`] = `[]`;

exports[`buildCellOutputs > handles HTML output 1`] = `
[
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"text/html","channel":"output","data":"<div>Hello <b>world</b></div>","timestamp":0},"consoleOutputs":[],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles JSON output 1`] = `
[
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"application/json","channel":"output","data":{"foo":"bar","count":42},"timestamp":0},"consoleOutputs":[],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles application/vnd.marimo+traceback output 1`] = `
[
  {
    "items": [
      {
        "data": "<b>Traceback (most recent call last):</b>
  File "<stdin>", line 1, in <module>
TypeError: invalid value",
        "mime": "text/html",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles boolean false output 1`] = `
[
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"application/json","channel":"output","data":false,"timestamp":0},"consoleOutputs":[],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles empty marimo error data array 1`] = `[]`;

exports[`buildCellOutputs > handles empty state 1`] = `[]`;

exports[`buildCellOutputs > handles marimo error output 1`] = `
[
  {
    "items": [
      {
        "data": "Syntax error: Invalid syntax",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "marimo-error",
    },
  },
]
`;

exports[`buildCellOutputs > handles mix of empty and non-empty console outputs 1`] = `
[
  {
    "items": [
      {
        "data": "Actual output
",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
  {
    "items": [
      {
        "data": "Actual error
",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
]
`;

exports[`buildCellOutputs > handles mixed output and console streams 1`] = `
[
  {
    "items": [
      {
        "data": "Processing...
",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
  {
    "items": [
      {
        "data": "Warning: deprecated function
",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"text/html","channel":"output","data":"<div>Result: 42</div>","timestamp":2},"consoleOutputs":[{"mimetype":"text/plain","channel":"stdout","data":"Processing...\\n","timestamp":0},{"mimetype":"text/plain","channel":"stderr","data":"Warning: deprecated function\\n","timestamp":1}],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles multiple console outputs 1`] = `
[
  {
    "items": [
      {
        "data": "Line 1
",
        "mime": "application/vnd.code.notebook.stdout",
      },
      {
        "data": "Line 2
",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
  {
    "items": [
      {
        "data": "Warning: something happened
",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
]
`;

exports[`buildCellOutputs > handles multiple errors in marimo error output 1`] = `
[
  {
    "items": [
      {
        "data": "ValueError: ValueError: invalid value",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "marimo-error",
    },
  },
]
`;

exports[`buildCellOutputs > handles multiple stderr errors 1`] = `
[
  {
    "items": [
      {
        "data": "Error 1: Connection failed
",
        "mime": "application/vnd.code.notebook.stderr",
      },
      {
        "data": "Error 2: Retry failed
",
        "mime": "application/vnd.code.notebook.stderr",
      },
      {
        "data": "Error 3: Timeout
",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
]
`;

exports[`buildCellOutputs > handles null output object 1`] = `[]`;

exports[`buildCellOutputs > handles numeric zero output 1`] = `
[
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"application/json","channel":"output","data":0,"timestamp":0},"consoleOutputs":[],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles stderr output 1`] = `
[
  {
    "items": [
      {
        "data": "Error message",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
]
`;

exports[`buildCellOutputs > handles stdin output 1`] = `
[
  {
    "items": [
      {
        "data": "Enter your name: ",
        "mime": "text/plain",
      },
    ],
    "metadata": {
      "channel": "stdin",
    },
  },
]
`;

exports[`buildCellOutputs > handles stdout + stderr + output together 1`] = `
[
  {
    "items": [
      {
        "data": "Starting computation...
",
        "mime": "application/vnd.code.notebook.stdout",
      },
      {
        "data": "Computation complete
",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
  {
    "items": [
      {
        "data": "Warning: using deprecated API
",
        "mime": "application/vnd.code.notebook.stderr",
      },
    ],
    "metadata": {
      "channel": "stderr",
    },
  },
  {
    "items": [
      {
        "data": "{"cellId":"test-cell-id","state":{"outline":null,"output":{"mimetype":"application/json","channel":"output","data":{"result":"success","value":100},"timestamp":3},"consoleOutputs":[{"mimetype":"text/plain","channel":"stdout","data":"Starting computation...\\n","timestamp":0},{"mimetype":"text/plain","channel":"stderr","data":"Warning: using deprecated API\\n","timestamp":1},{"mimetype":"text/plain","channel":"stdout","data":"Computation complete\\n","timestamp":2}],"status":"idle","staleInputs":false,"interrupted":false,"errored":false,"stopped":false,"runElapsedTimeMs":null,"runStartTimestamp":null,"lastRunStartTimestamp":null,"debuggerActive":false}}",
        "mime": "application/vnd.marimo.ui+json",
      },
    ],
    "metadata": undefined,
  },
]
`;

exports[`buildCellOutputs > handles stdout output 1`] = `
[
  {
    "items": [
      {
        "data": "Hello from stdout",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
]
`;

exports[`buildCellOutputs > preserves whitespace-only output 1`] = `
[
  {
    "items": [
      {
        "data": "   
",
        "mime": "application/vnd.code.notebook.stdout",
      },
    ],
    "metadata": {
      "channel": "stdout",
    },
  },
]
`;
